src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorEditLinesHelper.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * AceEditorEditLinesHelper.java
+ *
+ * Copyright (C) 2009-18 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.studio.client.workbench.views.source.editors.text;
+
+import org.rstudio.core.client.StringUtil;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.Position;
+import org.rstudio.studio.client.workbench.views.source.editors.text.ace.Range;
+import org.rstudio.studio.client.workbench.views.source.editors.text.events.CursorChangedEvent;
+import org.rstudio.studio.client.workbench.views.source.editors.text.events.CursorChangedHandler;
+
+import com.google.gwt.event.shared.HandlerRegistration;
+
+public class AceEditorEditLinesHelper implements CursorChangedHandler
+{
+   public AceEditorEditLinesHelper(AceEditor editor)
+   {
+      editor_ = editor;
+   }
+   
+   public void editLinesFromStart()
+   {
+      int state = state_;
+      state_ = (state_ + 1) % 4;
+      
+      if (cursorChangedHandler_ == null)
+         cursorChangedHandler_ = editor_.addCursorChangedHandler(this);
+      
+      if (range_ == null)
+         range_ = getSelectionExtent();
+      
+      isPerformingEditLinesAction_ = true;
+      
+      switch (state)
```

Comment:
Nit: Maybe a try/catch around this so that an exception can't stick `isPerformingEditLinesAction_` in the wrong state?